### PR TITLE
INTMDB 416 database user

### PR DESCRIPTION
**Pull request by ParthasarathyVaradhan**
## Description

Please include a summary of the fix/feature/change, including any relevant motivation and context.

Link to any related issue(s): 

## Type of change:

- [✅] Bug fix (non-breaking change which fixes an issue)
- [✅ ] New feature (non-breaking change which adds functionality)
-    [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [✅ ] This change requires a documentation update

## Required Checklist:

- [✅] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added any necessary documentation (if appropriate)
- [ ] I have run `make fmt` and formatted my code

## Further comments

INTMDB-416-database-user

New Change:
Code for new X509 database-user type.

Bug Fixes And Enhancements: 
Scopes updates during update operation.
Validation of input parameters.

Testing templates:
examples/database-user/user.json

### CODE REVIEW

File: cfn-resources/database-user/cmd/validation/validate.go [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+}
+func (m ModelValidator) GetUpdateFields() []string {
+	return UpdateRequiredFields
+}
+func (m ModelValidator) GetDeleteFields() []string {
+	return DeleteRequiredFields
+}
+func (m ModelValidator) GetListFields() []string {
+	return ListRequiredFields
+}
```
Comment (matt-condon) [inline cfn-resources/database-user/cmd/validation/validate.go:25]: Is it possible to reuse the existing [validator class](https://github.com/mongodb/mongodbatlas-cloudformation-resources/blob/master/cfn-resources/util/validator/validator.go)?

Comment (matt-condon) [reply cfn-resources/database-user/cmd/validation/validate.go:25]: If not, can we name them appropriately, and move this one to the same package

Comment (matt-condon) [reply cfn-resources/database-user/cmd/validation/validate.go:25]: Also, we  need unit tests to prevent breaking changes

Comment (ParthasarathyVaradhan) [reply cfn-resources/database-user/cmd/validation/validate.go]: Validator class is reused.
Validator class Changed in Util package.
Unit Test cases for Validator class are in place.

File: cfn-resources/database-user/cmd/resource/resource.go [hunk 2]
```diff
@@ -19,17 +17,23 @@ func setup() {
...
+// function to validate inputs to all actions
+func validateModel(fields []string, model *Model) *handler.ProgressEvent {
+	return validator.ValidateModel(fields, model)
 }
 
+var CreateRequiredFields = []string{"ApiKeys.PublicKey", "ApiKeys.PrivateKey", "DatabaseName", "ProjectId", "Roles", "Username"}
+var ReadRequiredFields = []string{"ApiKeys.PublicKey", "ApiKeys.PrivateKey", "ProjectId", "DatabaseName", "Username"}
+var UpdateRequiredFields = []string{"ApiKeys.PublicKey", "ApiKeys.PrivateKey", "DatabaseName", "ProjectId", "Roles", "Username"}
+var DeleteRequiredFields = []string{"ApiKeys.PublicKey", "ApiKeys.PrivateKey", "ProjectId", "DatabaseName", "Username"}
+var ListRequiredFields = []string{"ApiKeys.PublicKey", "ApiKeys.PrivateKey", "ProjectId"}
```
Comment (matt-condon) [inline cfn-resources/database-user/cmd/resource/resource.go:29]: Mentioned this in 
but we have almost the same logic in https://github.com/mongodb/mongodbatlas-cloudformation-resources/blob/master/cfn-resources/project/cmd/validation/validate.go

Would it be possible to reuse that?

Comment (ParthasarathyVaradhan) [reply cfn-resources/database-user/cmd/resource/resource.go:29]: We have different fields required for each methods in different resources.
For Example,
Required Fields for Read Method in Database-User: {"ApiKeys.PublicKey", "ApiKeys.PrivateKey", "ProjectId", **"DatabaseName", "Username"**}
Required Fields for Read Method in Cloud-Backup: {"ProjectId", **"ClusterName", "Id",** "ApiKeys.PublicKey", "ApiKeys.PrivateKey"}

Comment (matt-condon) [reply cfn-resources/database-user/cmd/resource/resource.go:29]: Cool, thank you for clarifying!

**Review by matt-condon — APPROVED** [review]

**Pull request closed without merging**